src/compiler/binder.ts

```diff
@@ -1308,6 +1308,7 @@ namespace ts {
                 case SyntaxKind.JSDocFunctionType:
                 case SyntaxKind.ModuleDeclaration:
                 case SyntaxKind.TypeAliasDeclaration:
+                case SyntaxKind.MappedType:

```

Comment:
This is because of the type parameter introduced by `[P in ...]`. We want that symbol entered into the `locals` symbol table such that it is accessible within the mapped type, but not from the outside (unlike members).